ovn-tester/ovn_workload.py

```diff
@@ -528,83 +548,90 @@ def __init__(self, central_node, worker_nodes, cluster_cfg, brex_cfg):
                               str(self.central_node.mgmt_ip + 2)]
         else:
             nb_cluster_ips = [str(self.central_node.mgmt_ip)]
-        self.nbctl.start_daemon(nb_cluster_ips, self.cluster_cfg.enable_ssl)
-        self.nbctl.set_global(
+        await self.nbctl.start_daemon(nb_cluster_ips,
+                                      self.cluster_cfg.enable_ssl)
+        await self.nbctl.set_global(
             'use_logical_dp_groups',
             self.cluster_cfg.logical_dp_groups
         )
-        self.nbctl.set_global(
+        await self.nbctl.set_global(
             'northd_probe_interval',
             self.cluster_cfg.northd_probe_interval
         )
-        self.nbctl.set_inactivity_probe(self.cluster_cfg.db_inactivity_probe)
-        self.sbctl.set_inactivity_probe(self.cluster_cfg.db_inactivity_probe)
+        await self.nbctl.set_inactivity_probe(
+            self.cluster_cfg.db_inactivity_probe
+        )
+        await self.sbctl.set_inactivity_probe(
+            self.cluster_cfg.db_inactivity_probe
+        )
 
-    def create_cluster_router(self, rtr_name):
-        self.router = self.nbctl.lr_add(rtr_name)
+    async def create_cluster_router(self, rtr_name):
+        self.router = await self.nbctl.lr_add(rtr_name)
 
-    def create_cluster_load_balancer(self, lb_name):
-        self.load_balancer = lb.OvnLoadBalancer(lb_name, self.nbctl,
-                                                self.cluster_cfg.vips)
-        self.load_balancer.add_vips(self.cluster_cfg.static_vips)
+    async def create_cluster_load_balancer(self, lb_name):
+        self.load_balancer = await create_load_balancer(lb_name, self.nbctl,
+                                                        self.cluster_cfg.vips)
+        await self.load_balancer.add_vips(self.cluster_cfg.static_vips)
 
-    def create_cluster_join_switch(self, sw_name):
-        self.join_switch = self.nbctl.ls_add(sw_name,
-                                             self.cluster_cfg.gw_net)
+    async def create_cluster_join_switch(self, sw_name):
+        self.join_switch = await self.nbctl.ls_add(sw_name,
+                                                   self.cluster_cfg.gw_net)
 
         lrp_ip = netaddr.IPAddress(self.cluster_cfg.gw_net.last - 1)
-        self.join_rp = self.nbctl.lr_port_add(
+        self.join_rp = await self.nbctl.lr_port_add(
             self.router, 'rtr-to-join', RandMac(), lrp_ip,
             self.cluster_cfg.gw_net.prefixlen
         )
-        self.join_ls_rp = self.nbctl.ls_port_add(
+        self.join_ls_rp = await self.nbctl.ls_port_add(
             self.join_switch, 'join-to-rtr', self.join_rp
         )
 
-    def provision_ports(self, n_ports, passive=False):
-        return [
-            self.select_worker_for_port().provision_ports(self, 1, passive)[0]
-            for _ in range(n_ports)
-        ]
+    async def provision_ports(self, n_ports, passive=False):
+        ret_list = []
+        for _ in range(n_ports):
```

Comment:
HI Mark these changes looks very good to me but here maybe we can create ports concurrently using asyncio.gather rather than await i think gather will consume less time if we create list of coroutines to create list of ports at the same time : (please see the suggested change below):

-        for _ in range(n_ports):
+        for _ in range(n_ports // len(self.worker_nodes)):
+            tasks = [fun(self, 1, passive) for fun in self.worker_nodes]
+            L = await asyncio.gather(*tasks)
+            ret_list.append([x[0] for x in L])
+        
+        for _ in range(n_ports % len(self.worker_nodes)):
             worker = self.select_worker_for_port()
             ports = await worker.provision_ports(self, 1, passive)
             ret_list.append(ports[0])

Reply:
Hi Mohammad. Thanks for the suggestion. I updated the PR with your suggestion. I added it on as a final commit. I did things a bit differently though. Rather than performing batches of provisioning based on the number of worker nodes, I instead grouped all of the provisioning into a single array and gathered that. It made the code a bit more readable IMO.

Reply:
Hi Mohammad. I implemented this suggestion, and then I tried it out on a test machine. Even though I can't see any issue with the code, it fails consistently for me.

I added some debug. With the original code, this is what I see happen:
```
2021-09-01 14:06:09,657 | ovn_workload |INFO| Creating localnet on ovn-scale-0
2021-09-01 14:06:09,657 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ovs-vsctl -- set open_vswitch . external-ids:ovn-bridge-mappings=providernet:br-ex"
2021-09-01 14:06:09,744 | ovn_workload |INFO| Adding external host on ovn-scale-0
2021-09-01 14:06:09,744 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ip link add veth0 type veth peer name veth1"
2021-09-01 14:06:09,838 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ip link add veth0 type veth peer name veth1"
2021-09-01 14:06:09,939 | ovn_sandbox  |WARNING| Command docker exec ovn-scale-0 ip link add veth0 type veth peer name veth1 exited with status 2
2021-09-01 14:06:09,939 | ovn_sandbox  |WARNING| stderr: RTNETLINK answers: File exists

2021-09-01 14:06:09,939 | ovn_sandbox  |WARNING| stdout: 
2021-09-01 14:06:09,939 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ip netns add ext-ns"
2021-09-01 14:06:10,039 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ip link set netns ext-ns dev veth0"
2021-09-01 14:06:10,142 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ip netns exec ext-ns ip link set dev veth0 up"
2021-09-01 14:06:10,236 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ip netns exec ext-ns ip addr add 3.0.255.253/16 dev veth0"
2021-09-01 14:06:10,339 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ip netns exec ext-ns ip route add default via 3.0.255.254"
2021-09-01 14:06:10,422 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ip link set dev veth1 up"
2021-09-01 14:06:10,524 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ovs-vsctl add-port br-ex veth1"
```

The `ovn-scale-0` container is created, and a bunch of commands to set up network interfaces is run. Each of those succeeds (with the benign error of adding the same veth0 interface twice).

When I switch to using `asyncio.gather()`, it appears the worker container is not fully started up before we start issuing network commands. Adding some debug, I see results like this:
```
2021-09-01 13:52:32,877 | ovn_workload |INFO| Starting worker ovn-scale-0
2021-09-01 13:52:32,877 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "cd /root/ovn-heater/runtime/ovn-fake-multinode && OVN_MONITOR_ALL=yes OVN_DB_CLUSTER=yes ENABLE_SSL=yes ENABLE_ETCD=no OVN_DP_TYPE=system CREATE_FAKE_VMS=no CHASSIS_COUNT=0 GW_COUNT=0 RELAY_COUNT=3 IP_HOST=192.16.0.0 IP_CIDR=16 IP_START=192.16.0.2 ./ovn_cluster.sh add-chassis ovn-scale-0 tcp:0.0.0.1:6642"
2021-09-01 13:52:32,878 | ovn_workload |INFO| Creating localnet on ovn-scale-0
2021-09-01 13:52:32,878 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ovs-vsctl -- set open_vswitch . external-ids:ovn-bridge-mappings=providernet:br-ex"
2021-09-01 13:52:32,879 | ovn_workload |INFO| Starting worker ovn-scale-1
2021-09-01 13:52:32,879 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "cd /root/ovn-heater/runtime/ovn-fake-multinode && OVN_MONITOR_ALL=yes OVN_DB_CLUSTER=yes ENABLE_SSL=yes ENABLE_ETCD=no OVN_DP_TYPE=system CREATE_FAKE_VMS=no CHASSIS_COUNT=0 GW_COUNT=0 RELAY_COUNT=3 IP_HOST=192.16.0.0 IP_CIDR=16 IP_START=192.16.0.3 ./ovn_cluster.sh add-chassis ovn-scale-1 tcp:0.0.0.1:6642"
2021-09-01 13:52:32,880 | ovn_workload |INFO| Creating localnet on ovn-scale-1
2021-09-01 13:52:32,880 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-1 ovs-vsctl -- set open_vswitch . external-ids:ovn-bridge-mappings=providernet:br-ex"
2021-09-01 13:52:32,938 | ovn_sandbox  |WARNING| Command docker exec ovn-scale-1 ovs-vsctl -- set open_vswitch . external-ids:ovn-bridge-mappings=providernet:br-ex exited with status 1
2021-09-01 13:52:32,938 | ovn_sandbox  |WARNING| stderr: Error: No such container: ovn-scale-1

2021-09-01 13:52:32,938 | ovn_sandbox  |WARNING| stdout: 
2021-09-01 13:52:32,938 | ovn_workload |INFO| Adding external host on ovn-scale-1
2021-09-01 13:52:32,938 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-1 ip link add veth0 type veth peer name veth1"
2021-09-01 13:52:32,939 | ovn_sandbox  |WARNING| Command docker exec ovn-scale-0 ovs-vsctl -- set open_vswitch . external-ids:ovn-bridge-mappings=providernet:br-ex exited with status 1
2021-09-01 13:52:32,939 | ovn_sandbox  |WARNING| stderr: Error: No such container: ovn-scale-0

2021-09-01 13:52:32,939 | ovn_sandbox  |WARNING| stdout: 
2021-09-01 13:52:32,939 | ovn_workload |INFO| Adding external host on ovn-scale-0
2021-09-01 13:52:32,939 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ip link add veth0 type veth peer name veth1"
2021-09-01 13:52:32,984 | ovn_sandbox  |WARNING| Command docker exec ovn-scale-0 ip link add veth0 type veth peer name veth1 exited with status 1
2021-09-01 13:52:32,984 | ovn_sandbox  |WARNING| stderr: Error: No such container: ovn-scale-0

2021-09-01 13:52:32,984 | ovn_sandbox  |WARNING| stdout: 
2021-09-01 13:52:32,984 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-0 ip link add veth0 type veth peer name veth1"
2021-09-01 13:52:32,985 | ovn_sandbox  |WARNING| Command docker exec ovn-scale-1 ip link add veth0 type veth peer name veth1 exited with status 1
2021-09-01 13:52:32,985 | ovn_sandbox  |WARNING| stderr: Error: No such container: ovn-scale-1

2021-09-01 13:52:32,985 | ovn_sandbox  |WARNING| stdout: 
2021-09-01 13:52:32,985 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-1 ip link add veth0 type veth peer name veth1"
2021-09-01 13:52:33,031 | ovn_sandbox  |WARNING| Command docker exec ovn-scale-1 ip link add veth0 type veth peer name veth1 exited with status 1
2021-09-01 13:52:33,031 | ovn_sandbox  |WARNING| stderr: Error: No such container: ovn-scale-1

2021-09-01 13:52:33,032 | ovn_sandbox  |WARNING| stdout: 
2021-09-01 13:52:33,032 | ovn_sandbox  |INFO| Logging command: ssh wsfd-netdev72.ntdv.lab.eng.bos.redhat.com "docker exec ovn-scale-1 ip netns add ext-ns"
2021-09-01 13:52:33,033 | ovn_sandbox  |WARNING| Command docker exec ovn-scale-0 ip link add veth0 type veth peer name veth1 exited with status 1
2021-09-01 13:52:33,033 | ovn_sandbox  |WARNING| stderr: Error: No such container: ovn-scale-0
...
```
As you can see, it now reports errors about the containers not existing. I'm guessing this is a temporary issue. By the time we move on in the tests and start trying to execute pings on those containers, we no longer get the "no such container" errors. However, since the network interfaces were not set up properly on those containers, it means that the pings fail.

I think we need a more consistent way of knowing the container in question is now up and running. We currently assume that if the SSH command returns, the container is ready. But that appears not to be the case.

As such, I'm going to revert this suggestion for the time being, pending a change for container readiness detection.